spec/unit/functions/lookup_spec.rb

```diff
@@ -1,733 +1,1253 @@
-              :type => :merge
+                {
+                  :value => { 'k1' => 'env_e1', 'k2' => 'module_e2', 'k3' => 'env_e3' },
+                  :event => :result,
+                  :merge => :deep,
+                  :options => { 'merge_hash_arrays' => true },
+                  :type => :merge,
+                  :branches => [
+                    {
+                      :key => 'abc::e',
+                      :event => :not_found,
+                      :type => :global,
+                      :name => :hiera
+                    },
+                    {
+                      :type => :data_provider,
+                      :name => 'Environment Data Provider (lookup version 5)',
+                      :branches => [
+                        {
+                          :type => :data_provider,
+                          :name => 'Legacy function "environment::data"',
+                          :key => 'abc::e',
+                          :value => { 'k1' => 'env_e1', 'k3' => 'env_e3' },
+                          :event => :found
+                        }
+                      ]
+                    },
+                    {
+                      :type => :data_provider,
+                      :name => 'Module "abc" Data Provider (lookup version 5)',
+                      :module => 'abc',
+                      :branches => [
+                        {
+                          :type => :data_provider,
+                          :name => 'Legacy function "abc::data"',
+                          :key => 'abc::e',
+                          :event => :found,
+                          :value => {
+                            'k1' => 'module_e1',
+                            'k2' => 'module_e2'
+                          }
+                        }
+                      ]
+                    }
+                  ]
+                }
+              ]
             }
           )
+        end
       end
-    end
 
-    it 'will explain that "lookup_options" is an invalid key' do
-      assemble_and_compile('${r}', "'abc::a'") do |scope|
-        lookup_invocation = Puppet::Pops::Lookup::Invocation.new(scope, {}, {}, true)
-        begin
-          Puppet::Pops::Lookup.lookup('lookup_options', nil, nil, false, nil, lookup_invocation)
-        rescue Puppet::Error
-        end
-        expect(lookup_invocation.explainer.explain).to eq(<<EOS)
+      it 'will explain that "lookup_options" is an invalid key' do
+        assemble_and_compile('${r}', "'abc::a'") do |scope|
+          lookup_invocation = Puppet::Pops::Lookup::Invocation.new(scope, {}, {}, true)
+          begin
+            Puppet::Pops::Lookup.lookup('lookup_options', nil, nil, false, nil, lookup_invocation)
+          rescue Puppet::Error
+          end
+          expect(lookup_invocation.explainer.explain).to eq(<<EOS)
 Invalid key "lookup_options"
 EOS
+        end
       end
-    end
 
-    it 'will explain that "lookup_options" is
```

Comment:
Looks like this, and the two following contexts (lookup_key, and data_dig) would benefit from using a common set of tests "behaves like lookup in general" wrt. env overrides, etc. That would make it easier to see commonality and differences. (Maybe also break up this long spec into several)